xds/internal/resolver/watch_service_test.go

```diff
@@ -16,375 +16,375 @@
-// canceled and restarted.
-func (s) TestServiceNotCancelRDSOnSameLDSUpdate(t *testing.T) {
-	serviceUpdateCh := testutils.NewChannel()
-	xdsC := fakeclient.NewClient()
-	cancelWatch := watchService(xdsC, targetStr, func(update serviceUpdate, err error) {
-		serviceUpdateCh.Send(serviceUpdateErr{u: update, err: err})
-	}, nil)
-	defer cancelWatch()
+	// Wait for no update from the resolver.
+	verifyNoUpdateFromResolver(ctx, t, tcc)
 
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer cancel()
-	waitForWatchListener(ctx, t, xdsC, targetStr)
-	xdsC.InvokeWatchListenerCallback(xdsresource.ListenerUpdate{RouteConfigName: routeStr}, nil)
-	waitForWatchRouteConfig(ctx, t, xdsC, routeStr)
-
-	wantUpdate := serviceUpdate{virtualHost: &xdsresource.VirtualHost{Domains: []string{"target"}, Routes: []*xdsresource.Route{{
-		Prefix:           newStringP(""),
-		WeightedClusters: map[string]xdsresource.WeightedCluster{cluster: {Weight: 1}}}},
-	}}
-	xdsC.InvokeWatchRouteConfigCallback("", xdsresource.RouteConfigUpdate{
-		VirtualHosts: []*xdsresource.VirtualHost{
-			{
-				Domains: []string{targetStr},
-				Routes:  []*xdsresource.Route{{Prefix: newStringP(""), WeightedClusters: map[string]xdsresource.WeightedCluster{cluster: {Weight: 1}}}},
-			},
-		},
-	}, nil)
-
-	if err := verifyServiceUpdate(ctx, serviceUpdateCh, wantUpdate); err != nil {
+	// Update the management server with the new route configuration resource.
+	resources.Routes = append(resources.Routes, e2e.DefaultRouteConfig(newTestRouteConfigName, defaultTestServiceName, defaultTestClusterName))
+	if err := mgmtServer.Update(ctx, resources); err != nil {
 		t.Fatal(err)
 	}
 
-	// Another LDS update with a the same RDS_name.
-	xdsC.InvokeWatchListenerCallback(xdsresource.ListenerUpdate{RouteConfigName: routeStr}, nil)
-	sCtx, sCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
-	defer sCancel()
-	if _, err := xdsC.WaitForCancelRouteConfigWatch(sCtx); err != context.DeadlineExceeded {
-		t.Fatalf("wait for cancel route watch failed: %v, want nil", err)
-	}
+	// Ensure update from the resolver.
+	verifyUpdateFromResolver(ctx, t, tcc, wantDefaultServiceConfig)
 }
 
-// TestServiceWatchInlineRDS covers the cases switching between:
-// - LDS update contains RDS name to watch
-// - LDS update contains inline RDS resource
-func (s) TestServiceWatchInlineRDS(t *testing.T) {
-	serviceUpdateCh := testutils.NewChannel()
-	xdsC := fakeclient.NewClient()
-	cancelWatch := watchService(xdsC, targetStr, func(update serviceUpdate, err error) {
-		serviceUpdateCh.Send(serviceUpdateErr{u: update, err: err})
-	}, nil)
-	defer cancelWatch()
-
+// Tests the case where the listener resource changes to contain an inline route
+// configuration and changes back to having a route configuration resource name.
+// Verifies that the expected xDS resource names are requested by the resolver
+// and the update pushed to the channel contains the expected service config.
```

Comment:
Change this comment to callback based API is how verification happens mechanically.